### DISCUSSION TITLE
How can one determine the number of iterations run by conjugate gradient (cg)?

### DISCUSSION THREAD

**Discussion by zaccharieramzi** [jax-ml/jax#13583] [Q&A] [answered]
So currently the implementation of cg in jax (`jax.scipy.sparse.linalg.cg`) only has a placeholder for information about the convergence, and it is `None`.

I would still like to know for how many iterations the solver has run: is there a hack that I could put in place, potentially modifying jax's source code, to get this information?

**Comment by zaccharieramzi** [comment]
Apparently there was actually an effort to do exactly that here: https://github.com/google/jax/pull/7825
but for some reason it has stalled, I'll try to see what I can make of it.

**Reply by zaccharieramzi** [reply]
and it's discussed here https://github.com/google/jax/issues/7909

**Comment by zaccharieramzi** [comment, marked as answer]
So what I got from #7825 , was the following: the information is contained in the output of the lax while loop which is actually discarded in the current cg solve implementation.
Therefore what one can do for a quick hack is to modify the `jax/scipy/sparse/linalg.py` ([here](https://github.com/google/jax/blob/main/jax/_src/scipy/sparse/linalg.py#L136)) and change the line to 

```python
x_final, r, *_, k = lax.while_loop(cond_fun, body_fun, initial_value)
```

You can then either print when not jitted or add `jax.debug.print('CG iterations: {k}', k=k)` when jitted (with `import jax` at the top of the file).

I want to highlight that I do not think it's a great idea to modify source code directly, but I couldn't use the PR code from #7825 directly + I had trouble running the tests locally so can't really take part in the PR.

**Comment by PhilipVinc** [comment]
You can use the solvers in this file https://gist.github.com/PhilipVinc/34e15d4fc93ee58753c24b29a8fba56e which won't require you to modify the source of jax and has been relatively stable in recent versions of jax (though it uses internal things so it might eventually break one day).
